app/common/metrics.py

```diff
@@ -9,3 +12,48 @@
     'build_state_duration_seconds',
     'Total amount of time a build spends in each build state',
     ['state'])
+
+serialized_build_time_seconds = Histogram(  # pylint: disable=no-value-for-parameter
+    'serialized_build_time_seconds',
+    'Total amount of time that would have been consumed by builds if all work was done serially')
+
+build_estimation_error = Histogram(  # pylint: disable=no-value-for-parameter
+    'build_estimation_error',
+    'For each build, the sum of the absolute value of the difference between the estimated and actual time for each atom to complete')
+
+
+class SlavesCollector(object):
+    """
+    Prometheus collector to collect the total number of alive/dead/idle slaves connected to the master.
+    collect() is called once each time prometheus scrapes the /metrics endpoint. This class ensures that
+    1. The list of slaves only gets iterated through once per scrape
+    2. A single slave is is not double counted in 2 states
+    """
+    def __init__(self, get_slaves: Callable[[], List['app.master.slave.Slave']]):
+        self._get_slaves = get_slaves
+
+    def collect(self):
+        active, idle, dead = 0, 0, 0
+        for slave in self._get_slaves():
+            if not slave.is_alive(use_cached=True):
```

Comment:
I'd want to set use_cached=False here.

The slave's dead/alive status is only updated whenever the master makes network calls to the slave. If a slave is working on a subjob, then the master will never contact that slave again until the slave makes an API call to the master indicating that it completed the subjob--which could be never, if the slave were to die while running a subjob.